File: sync-plugin/src/sync/handler/CreateStructMemberSyncHandler.cpp
#include "sync/handler/CreateStructMemberSyncHandler.h"
#include "sync/SyncManager.h"
#include "SyncPlugin.h"

#include "struct.hpp"

bool CreateStructMemberSyncHandler::ApplyUpdateImpl(StructMemberUpdateData* updateData)
{
	// Struct
	tid_t t = get_struc_id(updateData->structName.c_str());
	struc_t *s = get_struc(t);

	// Member
	opinfo_t ti;

	switch (updateData->memberType)
	{
	case StructMemberType::Struct:
		ti.tid = get_struc_id(updateData->targetStructName.c_str());
		break;

	case StructMemberType::String:
		ti.strtype = updateData->stringType;
		break;

	case StructMemberType::Offset:
		ti.ri = updateData->offsetRefInfo;
		break;
	}

	// Create
	if (updateData->syncType == SyncType::CreateStructMember)
		add_struc_member(s, updateData->memberName.c_str(), updateData->offset, updateData->flag, &ti, updateData->size);
	else
		set_member_type(s, updateData->offset, updateData->flag, &ti, updateData->size);

	return true;
}

bool CreateStructMemberSyncHandler::HandleNotification(IdaNotification& notification, StructMemberUpdateData* updateData)
{
	struc_t* pStruct = va_arg(notification.args, struc_t*);
	member_t* pMember = va_arg(notification.args, member_t*);

	updateData->structName = get_struc_name(pStruct->id).c_str();
		
	// Member
	updateData->memberName = get_member_name2(pMember->id).c_str();

	updateData->offset = pMember->unimem() ? 0 : pMember->soff;
	updateData->size = (uint64_t)(pMember->unimem() ? pMember->eoff : (pMember->eoff - pMember->soff));
	updateData->flag = (uint32_t)pMember->flag;

	// Value
	opinfo_t ti;
	opinfo_t* pti = retrieve_member_info(pMember, &ti);

	if (pti)
	{
		// Struct
		if (isStruct(pMember->flag))
		{
			updateData->memberType = StructMemberType::Struct;
			updateData->targetStructName = get_struc_name(ti.tid).c_str();
		}
		// String
		else if (isASCII(pMember->flag))
		{
			updateData->memberType = StructMemberType::String;
			updateData->stringType = ti.strtype;
		}
		// Offset
		else if (isOff0(pMember->flag) || isOff1(pMember->flag))
		{
			updateData->memberType = StructMemberType::Offset;
			updateData->offsetRefInfo = ti.ri;
		}
		// Enum (unsupported)
		else if (isEnum0(pMember->flag) || isEnum1(pMember->flag))
		{
			updateData->memberType = StructMemberType::Enum;
		}
	}
	else
	{
		// Data
		updateData->memberType = StructMemberType::Data;
	}

	// Update Type
	if (notification.code == idb_event::struc_member_created)
		updateData->syncType = SyncType::CreateStructMember;
	else
		updateData->syncType = SyncType::ChangeStructMember;

	return true;
}

void CreateStructMemberSyncHandler::DecodePacketImpl(StructMemberUpdateData* updateData, NetworkBufferT<BasePacket>* packet)
{
	updateData->structName = packet->ReadString();
	updateData->memberName = packet->ReadString();
	packet->Read(&updateData->memberType);

	packet->Read(&updateData->offset);
	packet->Read(&updateData->size);
	packet->Read(&updateData->flag);

	switch (updateData->memberType)
	{
	case StructMemberType::Struct:
		updateData->targetStructName = packet->ReadString();
		break;

	case StructMemberType::String:
		packet->Read(&updateData->stringType);
		break;

	case StructMemberType::Offset:
	{
		packet->Read(&updateData->offsetRefInfo.target);
		packet->Read(&updateData->offsetRefInfo.base);
		packet->Read(&updateData->offsetRefInfo.tdelta);
		packet->Read(&updateData->offsetRefInfo.flags);
	} break;
	}
}

void CreateStructMemberSyncHandler::EncodePacketImpl(NetworkBufferT<BasePacket>* packet, StructMemberUpdateData* updateData)
{
	packet->WriteString(updateData->structName.c_str());
	packet->WriteString(updateData->memberName.c_str());
	packet->Write(updateData->memberType);

	packet->Write(&updateData->offset);
	packet->Write(&updateData->size);
	packet->Write(&updateData->flag);

	switch (updateData->memberType)
	{
	case StructMemberType::Struct:
		packet->WriteString(updateData->targetStructName.c_str());
		break;

	case StructMemberType::String:
		packet->Write(&updateData->stringType);
		break;

	case StructMemberType::Offset:
	{
		packet->Write(&updateData->offsetRefInfo.target);
		packet->Write(&updateData->offsetRefInfo.base);
		packet->Write(&updateData->offsetRefInfo.tdelta);
		packet->Write(&updateData->offsetRefInfo.flags);
	} break;
	}
}

File: sync-server/src/sync/handler/CreateStructMemberSyncHandler.ts
import { Long } from 'mongodb';
import { IdbUpdate } from './../../database/IdbUpdate';
import { NetworkBuffer } from './../../network/NetworkBuffer';
import { IdbUpdatePacket } from './../../network/packets/IdbUpdatePacket';
import { BasePacket } from './../../network/packets/BasePacket';
import { NetworkClient } from './../../network/NetworkClient';
import { ISyncHandler, SyncType } from './../ISyncHandler';

export enum StructMemberType {
    Data,
    Struct,
    String,
    Enum,
    Offset
}

export class StructMemberUpdateData extends IdbUpdate {
    public structName: string;
    public memberName: string;
    public memberType: StructMemberType;

    public offset: Long;
    public size: Long;
    public flag: number;

    public targetStructName: string;
    public stringType: number;

    // Offset (Refinfo)
    public refinfoTarget: Long;
    public refinfoBase: Long;
    public refinfoDelta: Long;
    public refinfoFlags: number;
}

export class CreateStructMemberSyncHandler implements ISyncHandler {
    public syncType: SyncType = SyncType.CreateStructMember;

    public decodePacket(updateData: StructMemberUpdateData, packet: IdbUpdatePacket) {
        updateData.structName = packet.buffer.readString();
        updateData.memberName = packet.buffer.readString();
        updateData.memberType = packet.buffer.readUInt8();

        updateData.offset = packet.buffer.readUInt64();
        updateData.size = packet.buffer.readUInt64();
        updateData.flag = packet.buffer.readUInt32();

        switch (updateData.memberType) {
            case StructMemberType.Struct: {
                updateData.targetStructName = packet.buffer.readString();
                break;
            }

            case StructMemberType.String: {
                updateData.stringType = packet.buffer.readInt32();
                break;
            }

            case StructMemberType.Offset: {
                updateData.refinfoTarget = packet.buffer.readUInt64();
                updateData.refinfoBase = packet.buffer.readUInt64();
                updateData.refinfoDelta = packet.buffer.readUInt64();
                updateData.refinfoFlags = packet.buffer.readUInt32();
                break;
            }
        }
    }

    public encodePacket(packet: IdbUpdatePacket, updateData: StructMemberUpdateData) {
        packet.buffer.writeString(updateData.structName);
        packet.buffer.writeString(updateData.memberName);
        packet.buffer.writeUInt8(updateData.memberType);

        packet.buffer.writeUInt64(updateData.offset);
        packet.buffer.writeUInt64(updateData.size);
        packet.buffer.writeUInt32(updateData.flag);

        switch (updateData.memberType) {
            case StructMemberType.Struct: {
                packet.buffer.writeString(updateData.targetStructName);
                break;
            }

            case StructMemberType.String: {
                packet.buffer.writeUInt32(updateData.stringType);
                break;
            }

            case StructMemberType.Offset: {
                packet.buffer.writeUInt64(updateData.refinfoTarget);
                packet.buffer.writeUInt64(updateData.refinfoBase);
                packet.buffer.writeUInt64(updateData.refinfoDelta);
                packet.buffer.writeUInt32(updateData.refinfoFlags);
                break;
            }
        }
    }

    public getUniqueIdentifier(update: StructMemberUpdateData) {
        return null;
    }

    public updateToString(updateData: any): string {
        return null;
    }
}